src/prelude.jl

```diff
@@ -1,28 +1,36 @@
 const NANSAFE_MODE_ENABLED = @load_preference("nansafe_mode", false)
-const DEFAULT_CHUNK_THRESHOLD = @load_preference("default_chunk_threshold", 12)
+const DEFAULT_CHUNK_THRESHOLD = Val(@load_preference("default_chunk_threshold", 12))
 
 const AMBIGUOUS_TYPES = (AbstractFloat, Irrational, Integer, Rational, Real, RoundingMode)
 
 const UNARY_PREDICATES = Symbol[:isinf, :isnan, :isfinite, :iseven, :isodd, :isreal, :isinteger]
 
 struct Chunk{N} end
 
-const CHUNKS = [Chunk{i}() for i in 1:DEFAULT_CHUNK_THRESHOLD]
+const CHUNKS = ntuple(j -> Chunk{j}(), DEFAULT_CHUNK_THRESHOLD)
 
-function Chunk(input_length::Integer, threshold::Integer = DEFAULT_CHUNK_THRESHOLD)
-    N = pickchunksize(input_length, threshold)
-    0 < N <= DEFAULT_CHUNK_THRESHOLD && return CHUNKS[N]
+_getval(::Val{n}) where {n} = n
+
+function Chunk(::Val{input_length}, ::Val{threshold} = DEFAULT_CHUNK_THRESHOLD) where {input_length, threshold}
+    N = pickchunksize(Val(input_length), Val(threshold))
+    0 < N <= _getval(DEFAULT_CHUNK_THRESHOLD) && return CHUNKS[N]
     return Chunk{N}()
 end
 
-function Chunk(x::AbstractArray, threshold::Integer = DEFAULT_CHUNK_THRESHOLD)
+function Chunk(input_length::Integer, threshold::Integer = _getval(DEFAULT_CHUNK_THRESHOLD))
+    return Chunk(Val(input_length), Val(threshold))
+end
+
+function Chunk(x::AbstractArray, ::Val{threshold} = DEFAULT_CHUNK_THRESHOLD) where {threshold}
     return Chunk(length(x), threshold)
 end
 
+Chunk(x::AbstractArray, threshold::Integer) = Chunk(x, Val(threshold))
+
 # Constrained to `N <= threshold`, minimize (in order of priority):
 #   1. the number of chunks that need to be computed
 #   2. the number of "left over" perturbations in the final chunk
-function pickchunksize(input_length, threshold = DEFAULT_CHUNK_THRESHOLD)
+function pickchunksize(::Val{input_length}, ::Val{threshold} = DEFAULT_CHUNK_THRESHOLD) where {input_length, threshold}
```

Comment:
This is used in downstream packages, e.g., in SciML: https://github.com/search?q=org%3ASciML+pickchunksize&type=code So even though technically it might not breaking, it will affect many downstream packages. Similarly, `Chunk` is used by quite a few downstream packages, e.g., SciML or also LogDensityProblemsAD (and hence e.g. Turing indirectly).

Since the next release will be breaking though anyway, maybe that's not a major issue. I think it would be good to keep the API simple though, ie., to not support both integers and `Val`s.